### PR TITLE
Add assertValueAt(int, value) to TestObserver

### PR DESCRIPTION
**Pull request by NickFirmani**
I found myself frequently writing tests that want to check that an observable emitted a certain value second or third, but I don't care what the first or other values one, so `assertValues(Object...))` is more brittle.

This PR implements an `assertValueAt(index, value)` method, similar to `assertValue(value)`. Tests are also added.

### CODE REVIEW

**Comment by akarnokd** [comment]
There is an `assertValueAt(int, Predicate)` just below your changes, what's wrong with that? Also there could be an overload problem because of a `T` - `Predicate` conflict in Java 8+.

**Comment by NickFirmani** [comment]
I personally make heavy use of the `assertValue(T)` method, BaseConsumer#324, which also has a similar paired method `assertValue(Predicate<T>)`. Using that method keeps my tests cleaner, when I know the exact value I expect. It seems pretty verbose to me to implement a `Predicate` when the body will just be `value.equals(expected)`.

If this causes an overload problem, we'll have the same problem with `assertValue(T)` and `assertValue(Predicate<T>)` as well

File: src/test/java/io/reactivex/observers/TestObserverTest.java
Comment (artem-zinnatullin) [inline src/test/java/io/reactivex/observers/TestObserverTest.java]: Can you please compare something else than integers (String will be ok)? 

Silly code change in assert method can break its logic but pass this test because both index and value are `2` :)

**Review by artem-zinnatullin — CHANGES_REQUESTED** [review]
LGTM, I would use that!

**Comment by vanniktech** [comment]
We need to find some kind of policy for adding new API methods or not.

I'd say let's not get this in. Write a Kotlin extension functions 🙃 or use the `assertValueAt` with a Predicate and then define a test class that implements the interface and has a static factory method `Predicate<T> just(T)`.

If this gets in I also would like to have `mergeWith`, `concatWith`, `startWith` variants for `Single`, `Completable` and `Maybe` on `Observable` & `Flowable`.

**Comment by artem-zinnatullin** [comment]
I'm pretty sure this method will be much more popular than `assertValueAt(int, Predicate)` but that's just my guess

Mr ~~Negotiator~~ @vanniktech 😸 

For simple equals comparisons (which I believe are 95+% of asserts on values) this variant is definitely better than `assertValueAt(int, Predicate)`:

- It has nicer message and prints both expected and actual values including their classes (sometimes that saves tons of time)
- It requires less symbols to type `assertValueAt(3, expected)` even comparing with Kotlin `assertValueAt(3) { it == expected }`

**Comment by akarnokd** [comment]
Independent additions are individually evaluated based on the merits, value, quality and maintenance requirements they pose.

This PR is self contained and adds a convenience pair to an existing method. Plus, once merged in, there will be no need for further development on it.

In contrast, all those `mergeWith(Single)` and co are just shorthands for `mergeWith(Single.toObservable())`s and alike and the premise of "optimize later" actually adds burden to the maintainer(s) of RxJava.

**Comment by vanniktech** [comment]
Alright, you got me convinced now. If those shorthands were to be implemented one by one and also in an optimized way, would you consider PRs for that?

**Comment by akarnokd** [comment]
Yes.

File: src/main/java/io/reactivex/observers/BaseTestConsumer.java
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/BaseTestConsumer.java]: `Objects` is a Java 8 component but `ObjectHelper` is internal, thus maybe let's say "with respect to null-safe Object.equals()".

File: src/main/java/io/reactivex/observers/BaseTestConsumer.java
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/BaseTestConsumer.java]: `@since 2.1.3 - experimental`

File: src/main/java/io/reactivex/observers/BaseTestConsumer.java
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/BaseTestConsumer.java]: `@Experimental`

File: src/main/java/io/reactivex/observers/BaseTestConsumer.java
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/BaseTestConsumer.java]: Use `s` instead of `values.size()` here.

**Review by vanniktech — APPROVED** [review]

**Review by akarnokd — CHANGES_REQUESTED** [review]

**Comment by NickFirmani** [comment]
Thanks for the reviews!

File: src/test/java/io/reactivex/observers/TestObserverTest.java
Comment (NickFirmani) [reply src/test/java/io/reactivex/observers/TestObserverTest.java]: Good call.

File: src/main/java/io/reactivex/observers/BaseTestConsumer.java
Comment (NickFirmani) [reply src/main/java/io/reactivex/observers/BaseTestConsumer.java]: Fixed the other javadoc I copied that from too :)

File: src/main/java/io/reactivex/observers/BaseTestConsumer.java
Comment (NickFirmani) [reply src/main/java/io/reactivex/observers/BaseTestConsumer.java]: Nice catch.

**Review by akarnokd — APPROVED** [review]

**Comment by NickFirmani** [comment]
CI is failing due to 
`ERROR: JAVA_HOME is set to an invalid directory: /usr/lib/jvm/java-7-oracle`

**Comment by akarnokd** [comment]
There has been an image change on Travis CI and JDK 7 is no longer supported. See #5531.

**Comment by akarnokd** [comment]
Could you rebase the PR so the build settings can take effect?

**Comment by NickFirmani** [comment]
Thanks!

**Pull request merged**
